les/serverpool.go

```diff
@@ -17,904 +17,393 @@
-// needs to be removed and removes it from the parent pool too with a callback function.
-func (q *poolEntryQueue) setLatest(entry *poolEntry) {
-	if q.queue[entry.queueIdx] == entry {
-		delete(q.queue, entry.queueIdx)
+// updateWeight calculates the node weight and updates the nodeWeight field and the
+// hasValue flag. It also saves the node state if necessary.
+func (s *serverPool) updateWeight(node *enode.Node, totalValue float64, totalDialCost uint64) {
+	weight := uint64(totalValue * nodeWeightMul / float64(totalDialCost))
+	if weight >= nodeWeightThreshold {
+		s.ns.SetState(node, sfHasValue, nodestate.Flags{}, 0)
+		s.ns.SetField(node, sfiNodeWeight, weight)
 	} else {
-		if len(q.queue) == q.maxCnt {
-			e := q.fetchOldest()
-			q.remove(e)
-			q.removeFromPool(e)
-		}
-	}
-	entry.queueIdx = q.newPtr
-	q.queue[entry.queueIdx] = entry
-	q.newPtr++
+		s.ns.SetState(node, nodestate.Flags{}, sfHasValue, 0)
+		s.ns.SetField(node, sfiNodeWeight, nil)
+	}
+	s.ns.Persist(node) // saved if node history or hasValue changed
+}
+
+// redialWait returns the suggested waiting time before the node should be queried or
+// dialed again
+func (n *nodeHistory) redialWait() time.Duration {
+	return time.Duration(float64(minRedialWait) * n.waitFactor)
+}
+
+// setRedialWait calculates and sets the redialWait timeout based on the service value
+// and dial cost accumulated during the last session/attempt and in total.
+// The waiting time is raised exponentially if no service value has been received in order
+// to prevent dialing an unresponsive node frequently for a very long time just because it
+// was useful in the past. It can still be occasionally dialed though and once it provides
+// a significant amount of service value again its waiting time is quickly reduced or reset
+// to the minimum.
+// Note: node weight is also recalculated and updated by this function.
+func (s *serverPool) setRedialWait(node *enode.Node, addDialCost int64, waitStep float64) {
+	n, _ := s.ns.GetField(node, sfiNodeHistory).(nodeHistory)
+	sessionValue, totalValue := s.serviceValue(node)
+	totalDialCost := s.addDialCost(&n, addDialCost)
+
+	// if the current dial session has yielded at least the average value/dial cost ratio
+	// then waitFactor should be reset to 1 (the minimum value). If the session value
+	// is below average but still positive then waitFactor is limited to the ration of
+	// average / current service value. If the attempt was unsuccessful then waitFactor
+	// is raised exponentially without limitation.
+	// Note: dialCost is used in the formula below even if dial was not attempted at all
+	// because the pre-negotiation query did not return a positive result. In this case
+	// the ratio has no meaning anyway and waitFactor is always raised, though in smaller
+	// steps because queries are cheaper and therefore we can allow more failed attempts.
+	a := totalValue * dialCost
```

Comment:
Can we set a hard limitation for redial interval. In theory if the peer doesn't behave as the expected, we can refuse to dial it. However, my concern is if we have some bugs which always try to disconnect the peer, then the interval will be very high.

Reply:
Though it is increased exponentially with each dial attempt, as a function of elapsed time it increases linearly even if it is never accessible so I don't think there is a reason to hard limit it. Instead maybe we can discard it if it hasn't been seen in the last few months. I'll think about that.

Reply:
Now I realized you are right and it is indeed possible that `redialWait` goes very high quickly because the node can be selected by `FairMix` directly from the discovery iterator. So it is dialed often and still the waiting time is raised every time.

Reply:
fixed